### DISCUSSION TITLE
How to catch and handle exceptions with batches?

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#2188] [Q&A] [answered]
I'm struggling a bit to find the 'proper' way to do this.

## My Setup
I have Drift running in an isolate, and use a batch to perform multiple queries inserting/updating some data. Sometimes, the data causes a foreign key constraint to fail.

## My Question
How do I then catch this exception? In my stdout, I can see an unhandled SqliteException.
```
E/flutter (31412): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: SqliteException(787): FOREIGN KEY constraint failed, constraint failed (code 787)
E/flutter (31412):   Causing statement: INSERT INTO **rest of my query**
E/flutter (31412): #0      DriftCommunication.request (package:drift/src/remote/communication.dart:109:66)
E/flutter (31412): #1      _BaseExecutor.runBatched (package:drift/src/remote/client_impl.dart:68:10)
E/flutter (31412): #2      Batch._runWith (package:drift/src/runtime/api/batch.dart:210:10)
E/flutter (31412): #3      Batch._commit (package:drift/src/runtime/api/batch.dart:194:15)
E/flutter (31412): <asynchronous suspension>
```

## What I'm trying
```dart
try {
  await db.batch((batch) async {
    batch.insert(...);
    batch.customStatement(...);
  });
} on SqliteException catch(e) {
  // display exception to user etc.
}
```
But it seems that the exception is not caught when doing this. Any guidance is appreciated :)

**Comment by frankvollebregt** [comment, marked as answer]
I just figured it out! You can just use the `.onError` on the `db.batch`:
```dart
await db.batch((batch) async {
    batch.insert(...);
    batch.customStatement(...);
  }).onError((error, stackTrace) {
    // display exception to user etc.
  });
```

**Reply by simolus3** [reply]
Using `onError` is actually the same as an `try-await-catch` - the difference is that you now catch all exception types and previously only caught `SqliteException`s.

For remotes (e.g isolates), drift will actually throw a `DriftRemoteException` which is why you were unable to see the exception before. Something like #2169 will solve this, but it's a breaking change.

**Reply by frankvollebregt** [reply]
Ah, thank you for the clarification. I was already a bit confused, since I was convinced the try-catch worked for me in the past. Somehow having a single exception type to catch would be very useful in the future :)
